n26/api.py

```diff
@@ -295,6 +304,84 @@ def _do_request(self, method: str = GET, url: str = "/", params: dict = None,
         if len(response.content) > 0:
             return response.json()
 
+    def get_encryption_key(self, public_key: str = None) -> dict:
+        """
+        Receive public encryption key for the JSON containing the encrypted PIN
+        """
+        return self._do_request(GET, BASE_URL_DE + '/api/encryption/key', params={
+            'publicKey': public_key})
+
+    def encrypt_user_pin(self, pin: str):
+        """
+        Encrypts user PIN and prepares to send it in a format specified in the documentation
+
+        :return: base64 encoded JSON containing the encrypted PIN and its encryption key
+        """
+        # generate AES256 key and IV
+        random_password = Random.get_random_bytes(32)
+        salt = Random.get_random_bytes(16)
+        key = PBKDF2(random_password, salt, 32, count=1000000, hmac_hash_module=SHA512)
+        iv = Random.new().read(AES.block_size)
+        key64 = base64.b64encode(key)
+        iv64 = base64.b64encode(iv)
+        # encode the key and iv as a json string
+        aes_secret = {
+        'secretKey': key64,
+        'iv': iv64,
+        }
+        # json string has to be represented in byte form for encryption
+        # simplejson used instead of json since regular json can handle only strings, not base64 encoded data
+        unencrypted_aes_secret = bytes(simplejson.dumps(aes_secret), 'utf-8')
+        # Encrypt the secret JSON with RSA using the provided public key
+        public_key = self.get_encryption_key()
+        public_key_non64 = base64.b64decode(public_key['publicKey'])
+        public_key_object = RSA.importKey(public_key_non64)
+        public_key_cipher = PKCS1_v1_5.new(public_key_object)
+        encrypted_secret = public_key_cipher.encrypt(unencrypted_aes_secret)
+        encrypted_secret64 = base64.b64encode(encrypted_secret)
+        # Encrypt user's pin
+        private_key_cipher = AES.new(key=key, mode=AES.MODE_CBC, iv=iv)
+        # the pin has to be padded and transformed into bytes for a correct ecnryption format
+        encrypted_pin = private_key_cipher.encrypt(pad(bytes(pin, 'utf-8'), 16))
+        encrypted_pin64 = base64.b64encode(encrypted_pin)
+
+        return encrypted_secret64, encrypted_pin64
+
+    def create_transaction(self):
+        """
+        Creates a bank transfer order
+
+        All the input should be provided by the user via the command line
+        """
+        # Get all the necessary transfer information from user's input
+        iban = input("Please enter recipient's IBAN (spaces are allowed): ")
```

Comment:
Is there a reason those inputs can't be given as command arguments?
Although I like the idea of interactively entering those values, all current commands take all necessary values as command arguments, not interactively. The only exception to that is the 6 digit SMS OTP code approval, since it has to be interactive.

Reply:
I understand that, and I agree. Lets keep this interactive for now and think about an argument based solution at a later stage.

However, I would still like to move all inputs and input validation logic to the `cli.py` class method, to separate the frontend from the backend. The same needs to be done for the initial login someday, but thats a topic for another day.